server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -1238,6 +1239,28 @@ private void executePipelines(
                         return; // document failed!
                     }
 
+                    for (StreamType streamType : StreamType.getEnabledStreamTypesForProject(project)) {
+                        if (streamType.matchesStreamPrefix(newIndex)
+                            && ingestDocument.getIndexHistory().contains(streamType.getStreamName()) == false) {
+                            exceptionHandler.accept(
+                                new IngestPipelineException(
+                                    pipelineId,
+                                    new IllegalArgumentException(
+                                        format(
+                                            "Pipelines can't re-route documents to child streams, but pipeline [%s] tried to reroute "
+                                                + "this document from index [%s] to index [%s]. Reroute history: %s",
+                                            pipelineId,
+                                            originalIndex,
+                                            newIndex,
+                                            String.join(" -> ", ingestDocument.getIndexHistory())
```

Comment:
Nit: I like the style of the arrow separator, but I think a comma separated list is more aligned with our log message style and perhaps a tad easier to parse if ever needed.

```suggestion
                                            String.join(", ", ingestDocument.getIndexHistory())
```